### PR TITLE
add sender/receiver support to set_difference

### PR DESCRIPTION
**Pull request by mooofin**
Part 1/4 for the set-operation family (issue #6535). set_intersection,set_union, set_symmetric_difference.

Adds S/R support for set_difference:

- Widen `set_operation` and `set_difference::parallel()` return types
  to `decltype(auto)` so the scheduler-executor sender passes through
- Drop empty-range early returns from `parallel()`; `set_operation`
  handles them via sentinel-initialized `set_chunk_data`
- New sender tests + CMakeLists.txt 

 set_intersection, set_union,set_symmetric_difference to be followed as separate PRs.

### CODE REVIEW

File: libs/core/algorithms/include/hpx/parallel/algorithms/set_difference.hpp [hunk 1]
```diff
@@ -260,32 +260,21 @@ namespace hpx::parallel {
...
-                {
-                    return detail::copy<result_type>().call(
-                        HPX_FORWARD(ExPolicy, policy), first1, last1, dest);
-                }
 
+                // Edge cases (first1==last1, first2==last2) are handled inside
+                // set_operation: set_chunk_data initializes to sentinel values
+                // so empty range1 produces {first1,first2,dest} naturally, and
+                // sequential_set_difference with empty range2 copies range1.
                 using buffer_type = typename set_operations_buffer<Iter3>::type;
```
[the remark this answers is not in the document]
Comment (mooofin) [reply libs/core/algorithms/include/hpx/parallel/algorithms/set_difference.hpp]: It's only used inside `!last_partition` branches, so fixed by using `first1[last_partition ? end1-1 : end1]` as a safe fallback index. Pushed in the latest commit.

**Comment by mooofin** [comment]
Traced the macos failing to  `HPX_CXX_CORE_EXPORT` on the `set_operation` function template on clang with `-fvisibility=hidden` this marks `index_queue_bulk_state` vtables as default-visibility, so the linker expects them from the dylib  (GCC is lenient, hence local build passed). will push shortly

File: libs/core/algorithms/include/hpx/parallel/algorithms/detail/set_operation.hpp [hunk 2]
```diff
@@ -243,25 +243,50 @@ namespace hpx::parallel::detail {
...
-                            dest + ch->start_index);
-                    },
-                    [](set_chunk_data* last) -> set_chunk_data* {
-                        return last;
-                    });
+            // In the scheduler-executor (S/R) path, foreach_partitioner<par>
+            // would call parallel_executor::bulk_sync_execute which creates
+            // index_queue_bulk_state<true,...> with a TU-local lambda type
+            // whose vtable destructor cannot be resolved from the shared
+            // library. Use a sequential loop in that path instead.
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/detail/set_operation.hpp]: Can you explain this in more detail, please? I don't think that lamdas have a vtable. Also `foreach_partitioner<par>` would call `parallel_executor::bulk_sync_execute` only under certain circumstances, and even then it doesn't mean that the chunks are being executed sequentially, only that the executor synchronizes before returning.

File: libs/core/algorithms/include/hpx/parallel/algorithms/set_difference.hpp [hunk 3]
```diff
@@ -260,32 +260,21 @@ namespace hpx::parallel {
...
-                {
-                    return result::get(
-                        result_type{HPX_MOVE(first1), HPX_MOVE(dest)});
-                }
-
-                if (first2 == last2)
-                {
-                    return detail::copy<result_type>().call(
-                        HPX_FORWARD(ExPolicy, policy), first1, last1, dest);
-                }
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/set_difference.hpp:287]: I think we should remove these edge case tests only for sender&receiver executors, not in the general case.

Comment (mooofin) [reply libs/core/algorithms/include/hpx/parallel/algorithms/set_difference.hpp:287]: i'll restore the edge-case tests for the general executors and only skip them on the sender/receiver executrs. Will push that change

File: libs/core/algorithms/include/hpx/parallel/algorithms/detail/set_operation.hpp (hunk 2 above)
Comment (mooofin) [reply libs/core/algorithms/include/hpx/parallel/algorithms/detail/set_operation.hpp]: i'll remove the comment and post the real root cause before merge ,the probelem   is a link-time failure in the scheduler-executor path when the copy goes through foreach_partitioner<par>, will dig into what's really unresolved there

File: libs/core/algorithms/include/hpx/parallel/algorithms/detail/set_operation.hpp [hunk 4]
```diff
@@ -243,11 +243,12 @@ namespace hpx::parallel::detail {
...
-            // would call parallel_executor::bulk_sync_execute which creates
-            // index_queue_bulk_state<true,...> with a TU-local lambda type
-            // whose vtable destructor cannot be resolved from the shared
-            // library. Use a sequential loop in that path instead.
+            //
+            // Scheduler-executor (S/R) policies copy sequentially: the
+            // parallel path would instantiate index_queue_bulk_state whose
+            // D0/D1 destructors are homed to libhpx_core and are missing for
+            // this policy, breaking the link under clang -fvisibility=hidden.
+            // See #7344.
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/detail/set_operation.hpp:251]: Which symbols are not visible for you concretely? I'd rather fix that than to force this algorithm into sequential execution.

**Comment by mooofin** [comment]
The undefined symbols (from nm, clang -fvisibility=hidden) were D0Ev/D1Ev of index_queue_bulk_state<true, parallel_{,unsequenced_}task_policy_shim<explicit_scheduler_executor<...>, default_parameters>, ...>.

Root cause: index_queue_bulk_state inherits from future_data<void> which inherits from future_data_base<future_data_void> (HPX_CORE_EXPORT, key function in libhpx_core). Clang homes D0/D1 of derived classes to the same TU as the base's key function. Since this template is parameterized on user scheduler/executor types, no instantiation is ever built into libhpx_core, so D0/D1 resolve nowhere. GCC emits them weak-local, so it passes.

Fix I have in mind: add __attribute__((visibility("default"))) to index_queue_bulk_state and index_queue_bulk_state_result, guarded with #if !defined(HPX_WINDOWS)  same pattern as reinitializable_static. Then revert the sequential workaround.

**Comment by hkaiser** [comment]
> Fix I have in mind: add **attribute**((visibility("default"))) to index_queue_bulk_state and index_queue_bulk_state_result, guarded with #if !defined(HPX_WINDOWS) same pattern as reinitializable_static. Then revert the sequential workaround.

You can use `HPX_ALWAYS_EXPORT` for this purpose. There is no need to wrap it into a `#if`/`#endif`.

**Comment by hkaiser** [comment]
> > libs/core/algorithms/tests/unit/algorithms/set_difference_sender.cpp (1)> `113-163`: _🎯 Functional Correctness_ | _🔵 Trivial_ | _⚡ Quick win_
> > > **Add a "both ranges empty" case.**
> > > The empty-range test covers "empty first" and "empty second" individually but not both empty simultaneously — the most degenerate case for the sentinel-initialized `set_chunk_data` handling this PR relies on.
> > > 
> > > Proposed additional case
> > > ```diff
> > >      // empty second range -> result equals first range
> > >      {
> > >          std::vector<std::size_t> c1 = test::random_fill(10007);
> > >          std::sort(std::begin(c1), std::end(c1));
> > >          std::vector<std::size_t> c2;
> > >          std::vector<std::size_t> c3(c1.size());
> > >  
> > >          auto result = tt::sync_wait(
> > >              ex::just(iterator(std::begin(c1)), iterator(std::end(c1)),
> > >                  std::begin(c2), std::end(c2), std::begin(c3)) |
> > >              hpx::set_difference(ex_policy.on(exec)));
> > >  
> > >          auto const result_last = hpx::get<0>(result.value());
> > >          HPX_TEST(result_last == std::begin(c3) + c1.size());
> > >          HPX_TEST(std::equal(std::begin(c3), result_last, std::begin(c1)));
> > >      }
> > > +
> > > +    // both ranges empty -> empty result, dest unchanged
> > > +    {
> > > +        std::vector<std::size_t> c1;
> > > +        std::vector<std::size_t> c2;
> > > +        std::vector<std::size_t> c3(1, static_cast<std::size_t>(-1));
> > > +
> > > +        auto result = tt::sync_wait(
> > > +            ex::just(iterator(std::begin(c1)), iterator(std::end(c1)),
> > > +                std::begin(c2), std::end(c2), std::begin(c3)) |
> > > +            hpx::set_difference(ex_policy.on(exec)));
> > > +
> > > +        auto const result_last = hpx::get<0>(result.value());
> > > +        HPX_TEST(result_last == std::begin(c3));
> > > +    }
> > > ```

This comment makes perfect sense.

**Comment by mooofin** [comment]
> > Fix I have in mind: add **attribute**((visibility("default"))) to index_queue_bulk_state and index_queue_bulk_state_result, guarded with #if !defined(HPX_WINDOWS) same pattern as reinitializable_static. Then revert the sequential workaround.
> 
> You can use `HPX_ALWAYS_EXPORT` for this purpose. There is no need to wrap it into a `#if`/`#endif`.

On the visibility fix: applying HPX_ALWAYS_EXPORT to index_queue_bulk_state / index_queue_bulk_state_result and reverting the sequential-copy workaround touches libs/core/executors .

Prefer I land it here and revert the sequential workaround, or as its own PR with the workaround staying until it merges?  I've added the "both ranges empty" test case as suggested

**Comment by hkaiser** [comment]
> Prefer I land it here and revert the sequential workaround, or as its own PR with the workaround staying until it merges? I've added the "both ranges empty" test case as suggested

I think having it here would be preferrable.

**Comment by mooofin** [comment]
I tried HPX_ALWAYS_EXPORT on both bulk-state structs, but clang++-21 still hits the original link error:
undefined reference to index_queue_bulk_state<true, ...>::~index_queue_bulk_state() [D0Ev/D1Ev]
(job log)(job: https://github.com/TheHPXProject/hpx/actions/runs/28777768742/job/85333250546)

Looks like the failing specialization only gets instantiated in the leaf TU (its type embeds the whole set_operation chain), so libhpx_core never emits its structors  the export attribute doesn't reach it, and being leaf-local I couldn't see a way to anchor it in executors either.

It also only shows up for set_operation the one algorithm nesting a parallel foreach_partitioner in its S/R continuation (f2); other .on(exec) sender tests all pass on clang-21. So it seems scoped to the set family, not a general executors issue...

Would you prefer (a) the sequential copy on the S/R path (only the cores-chunk stitch goes serial, f1 stays parallel), or (b) a different approach to the structor emission?  
PS (MSVC C2059 + clang-22 failures look unrelated  came in with the rebase.)

**Comment by hkaiser** [comment]
> Would you prefer (a) the sequential copy on the S/R path (only the cores-chunk stitch goes serial, f1 stays parallel), or (b) a different approach to the structor emission? 

let's do (a)

> PS (MSVC C2059 + clang-22 failures look unrelated came in with the rebase.)

This is caused by the `HPX_ALWAYS_EXPORT`, I believe.

**Comment by hkaiser** [comment]
> > libs/core/algorithms/include/hpx/parallel/algorithms/detail/set_operation.hpp (1)> `246-290`: _📐 Maintainability & Code Quality_ | _🔵 Trivial_ | _⚡ Quick win_
> > > **Duplicated copy/sentinel-check logic between the two branches.**
> > > Both the sequential loop (lines 255-267) and the `foreach_partitioner` lambda (lines 274-286) repeat the identical uninit-sentinel check and `std::copy` call. Extracting a shared lambda would avoid future drift between the two paths.

What about this comment?

**Comment by mooofin** [comment]
> > > libs/core/algorithms/include/hpx/parallel/algorithms/detail/set_operation.hpp (1)> `246-290`: _📐 Maintainability & Code Quality_ | _🔵 Trivial_ | _⚡ Quick win_
> > > > **Duplicated copy/sentinel-check logic between the two branches.**
> > > > Both the sequential loop (lines 255-267) and the `foreach_partitioner` lambda (lines 274-286) repeat the identical uninit-sentinel check and `std::copy` call. Extracting a shared lambda would avoid future drift between the two paths.
> 
> What about this comment?

 Done

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!
